### PR TITLE
Fix block sync problem

### PR DESCRIPTION
**Pull request by zhxymh**

### CODE REVIEW

File: src/AElf.OS/Jobs/BlockSyncJob.cs [hunk 1]
```diff
@@ -15,22 +17,27 @@ namespace AElf.OS.Jobs
...
+        private readonly IAnnouncementCacheProvider _announcementCacheProvider;
 
         public ILogger<BlockSyncJob> Logger { get; set; }
 
         public BlockSyncJob(IBlockAttachService blockAttachService,
             IOptionsSnapshot<NetworkOptions> networkOptions,
             IBlockchainService blockchainService,
             INetworkService networkService,
-            ITaskQueueManager taskQueueManager)
+            ITaskQueueManager taskQueueManager,
```
Comment (loning) [inline src/AElf.OS/Jobs/BlockSyncJob.cs]: Too many injections in BlockSyncJob, maybe we should implement some logic in the service layer or domain layer.

File: src/AElf.OS/Jobs/BlockSyncJob.cs [hunk 2]
```diff
@@ -7,6 +7,8 @@
 using AElf.Kernel.SmartContractExecution.Application;
 using AElf.OS.Network;
 using AElf.OS.Network.Application;
+using AElf.OS.Network.Infrastructure;
```
Comment (loning) [inline src/AElf.OS/Jobs/BlockSyncJob.cs]: should not in AElf.OS namespace to reference Network's Infrastructure layer

**Comment by loning** [comment]
@shiwk @rosona also, reviewers need to check the rules

File: src/AElf.OS/BlockSync/Application/IBlockDownloadService.cs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                    throw new InvalidOperationException(
+                        $"Previous block not match previous {lastDownloadBlockHash}, network back {blocks.First().Header.PreviousBlockHash}");
+                }
+
+                foreach (var block in blocks)
+                {
+                    Logger.LogDebug(
+                        $"Processing block {block},  longest chain hash: {chain.LongestChainHash}, best chain hash : {chain.BestChainHash}");
+
+                    var enqueueTimestamp = Timestamp.FromDateTime(DateTime.UtcNow);
```
Comment (loning) [inline src/AElf.OS/BlockSync/Application/IBlockDownloadService.cs]: Timestamp.FromDateTime(DateTime.UtcNow) =>
TimestampHelper.GetUtcNow()

File: src/AElf.OS/BlockSync/Application/IBlockSyncService.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+            var chain = await _blockchainService.GetChainAsync();
+            if (blockHash != null && blockHeight < chain.BestChainHeight + 5)
+            {
+                await _blockFetchService.FetchBlockAsync(blockHash, blockHeight, suggestedPeerPubKey);
+            }
+            else
+            {
+                if (_blockSyncStateProvider.BlockSyncJobEnqueueTime != null
+                    && DateTime.UtcNow - _blockSyncStateProvider.BlockSyncJobEnqueueTime.ToDateTime() >
```
Comment (loning) [inline src/AElf.OS/BlockSync/Application/IBlockSyncService.cs]: DateTime.UtcNow ???
Timestamp - Timestamp

**Comment by loning** [comment]
1. add retry logic to grpc https://github.com/grpc/grpc/issues/15623#issuecomment-398487843
2. fetch blocks from multiple n nodes, save the fast one, and random fetch from n - 1 nodes, save the fast one, and continue ...
3. save blocks in one queue, and process block in another queue.

**Comment by loning** [comment]
test this branch online until it works, solve this problem before 23:59 pm.

**Comment by loning** [comment]
switch online test environment to this branch

**Comment by zhxymh** [comment]
> switch online test environment to this branch

We are already using this branch.

File: src/AElf.OS.Core/Network/Application/NetworkService.cs [hunk 5]
```diff
@@ -97,113 +119,149 @@ public async Task<int> BroadcastTransactionAsync(Transaction tx)
...
+            else if (bestPeer.PubKey != peerPubKey)
+                peers.Add(bestPeer);
+            
+            Random rnd = new Random();
+            
+            // Fill with random peers.
+            List<IPeer> randomPeers = _peerPool.GetPeers()
+                .Where(p => p.PubKey != peerPubKey && (bestPeer == null || p.PubKey != bestPeer.PubKey))
+                .OrderBy(x => rnd.Next())
+                .Take(3 - peers.Count)
```
Comment (loning) [inline src/AElf.OS.Core/Network/Application/NetworkService.cs]: if peers.Count < 3 ??

please do more checks and think more

File: src/AElf.OS.Network.Grpc/GrpcPeer.cs [hunk 6]
```diff
@@ -27,96 +41,199 @@ public bool IsReady
...
         {
             var blockRequest = new BlocksRequest {PreviousBlockHash = firstHash, Count = count};
+            var blockInfo = $"{{ first: {firstHash}, count: {count} }}";
+            
+            GrpcRequest request = new GrpcRequest
+            {
+                ErrorMessage = $"Get blocks for {blockInfo} failed.",
+                MetricName = nameof(MetricNames.GetBlocks),
+                MetricInfo = $"Get blocks for {blockInfo}",
+                Timeout = 500
```
Comment (rosona) [inline src/AElf.OS.Network.Grpc/GrpcPeer.cs]: mv timeout to GrpcConstants

File: src/AElf.OS.Network.Grpc/GrpcServerService.cs [hunk 7]
```diff
@@ -97,23 +97,32 @@ public override async Task<ConnectReply> Connect(Handshake handshake, ServerCall
...
             {
                 var c = channel.WaitForStateChangedAsync(channel.State);
             }
+
+            var connectionInfo = new GrpcPeerInfo
+            {
+                PublicKey = pubKey,
+                PeerIpAddress = peerAddress,
+                ProtocolVersion = handshake.HskData.Version,
+                ConnectionTime = DateTime.UtcNow.ToTimestamp().Seconds,
```
Comment (rosona) [inline src/AElf.OS.Network.Grpc/GrpcServerService.cs]: Use TimestampHelper

File: src/AElf.OS.Network.Grpc/GrpcPeerPool.cs [hunk 8]
```diff
@@ -106,9 +114,18 @@ private async Task<bool> DialAsync(string ipAddress)
             }
 
             var pubKey = connectReply.Handshake.HskData.PublicKey.ToHex();
+            
+            var connectionInfo = new GrpcPeerInfo
+            {
+                PublicKey = pubKey,
+                PeerIpAddress = ipAddress,
+                ProtocolVersion = connectReply.Handshake.HskData.Version,
+                ConnectionTime = DateTime.UtcNow.ToTimestamp().Seconds,
```
Comment (rosona) [inline src/AElf.OS.Network.Grpc/GrpcPeerPool.cs]: Use TimestampHelper

File: src/AElf.OS.Core/Network/Application/NetworkService.cs [hunk 9]
```diff
@@ -97,113 +117,145 @@ public async Task<int> BroadcastTransactionAsync(Transaction tx)
...
             Logger.LogDebug($"Getting block by hash, hash: {hash} from {peer}.");
-            return await GetBlockAsync(hash, peer, tryOthersIfSpecifiedFails);
+            return await GetBlockAsync(hash, peer);
         }
 
-        private async Task<BlockWithTransactions> GetBlockAsync(Hash hash, string peer = null, bool tryOthersIfSpecifiedFails = false)
+        private async Task<BlockWithTransactions> GetBlockAsync(Hash hash, string peer = null)
+        {
+            var peers = SelectPeers(peer);
+            var block = await RequestBlockToAsync(hash, peers, peer);
```
Comment (rosona) [inline src/AElf.OS.Core/Network/Application/NetworkService.cs]: rename?

File: src/AElf.OS.Core/Network/Types/RequestMetric.cs [hunk 10]
```diff
@@ -0,0 +1,12 @@
+using System;
+
+namespace AElf.OS.Network
+{
+    public class RequestMetric
+    {
+        public long RoundTripTime { get; set; }
+        public string MethodName { get; set; }
+        public string Info { get; set; }
+        public DateTime RequestTime { get; set; }
```
Comment (rosona) [inline src/AElf.OS.Core/Network/Types/RequestMetric.cs]: change to Timestamp

File: src/AElf.OS.Core/Network/Application/NetworkService.cs (hunk 9 above)
Comment (samfrb) [reply src/AElf.OS.Core/Network/Application/NetworkService.cs]: Merged request blocks methods.

**Pull request merged**